[CMS-95] Create atomic component - Link - CMS

levadadenys opened the pull request:
<!--
  A summary of the change, including any relevant background, motivation, and context.
  If relevant, include a description, screenshots, and/or video of the existing and new behavior.
-->

## [CMS-95] Create atomic component - Link - CMS

* feat(link): added link/index.css to package exports
* feat(layout): connected @code-dot-org/component-library-styles/colors.scss to marketing/src/app/layout.tsx
* feat(marketing/link): added Link component
* feat(common/definitions): added common definitions
* feat(common/styles): added common styles
* feat(common/types): added SpacingNoneToM, SpacingNoneToL types


https://github.com/user-attachments/assets/188b45ef-e51b-4f18-b594-2bb5c4a1586e



## Links
- jira ticket: [CMS-95](https://codedotorg.atlassian.net/browse/CMS-95)

<!--
  Links to relevant external resources; ie, specification documents, Jira tickets, related PRs, Honeybadger errors, etc.
-->

<!--
- spec: []()
-->

## Testing story

<!--
  Does your change include appropriate tests?
  If so, please describe how the tests included in this PR are sufficient.
  If not, please explain why this change does not need to be tested.
-->

<!-- Other aspects to consider. Delete any sections that are not relevant to your change. -->

## Deployment strategy

## Follow-up work

<!--
  List (ideally with Jira links) any clean-up or technical debt that will be addressed in future work.
-->

## Privacy

<!--
  1.	Does this change involve the collection, use, or sharing of new Personal Data?
  2.	Does this change involve a new or changed use or sharing of existing Personal Data?
-->

## Security

<!-- Link to Jira task(s) where sensitive security issues are discussed privately. -->

## Caching

## PR Checklist:

<!--
  The final step! Before you create your PR, double-check that everything is in order.
  Change [ ] to [X] during creation to check boxes.
-->

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

moshebaricdo commented on frontend/apps/marketing/src/components/link/LinkContentfulDefinition.ts:
```diff
@@ -0,0 +1,81 @@
+    'https://images.ctfassets.net/90t6bu6vlf76/2CPKrKCB3KxD1n6wG9JTn9/aab22373a39e9cc5305b21c08bba588d/component_link_thumbnail.png',
+  tooltip: {
+    description:
+      'A standalone text that directs users to internal or external pages, with options for styling and accessibility.',
```
This is tiny, but let's add "link" after "a standalone text."

levadadenys replied:
updated

moshebaricdo reviewed (commented):
Left a tiny text edit for help text in Studio. Otherwise everything LGTM! Will leave as comment so an eng can give approval!

kelbyhawn commented on frontend/apps/marketing/src/components/link/Link.tsx:
```diff
@@ -0,0 +1,57 @@
+  size: ComponentSizeXSToL;
+  /** Whether Link is for internal code.org pages, or external web page. (external links are opened in new tab) */
+  isLinkExternal: boolean;
+  /** ClassName passed by contentful to apply styles that are set through contentful native editor*/
```
Super nit:
```suggestion
  /** ClassName passed by Contentful to apply styles that are set through Contentful native editor */
```

kelbyhawn commented on frontend/apps/marketing/src/components/link/Link.tsx:
```diff
@@ -0,0 +1,57 @@
+  /** Link color */
+  color: 'purple' | 'white';
+  /** Link vertical margin */
+  bottomMargin: 'none' | 'xs' | 's' | 'm';
```
Semi-related consistency question: I'm using `ComponentSizeXSToL` for padding on the `Section` component ([ref](https://github.com/code-dot-org/code-dot-org/blob/aac72111896632f25701a11533cbabf688c07279/frontend/packages/component-library/src/cms/section/Section.tsx#L28)), should I remove that there and just use explicit values as seen here, or should you use the `ComponentSizeXSToL` here with the excluded values? 

This was also a change requested on the `Divider` component ([ref](https://github.com/code-dot-org/code-dot-org/blob/aac72111896632f25701a11533cbabf688c07279/frontend/packages/component-library/src/divider/Divider.tsx#L11)) so I'd like to norm around this either way.

levadadenys replied:
You got me 😄 
And you're right, we should also use `ComponentSizeXSToL` or something similar for Link and all the other components. Will update

Especially now that it looks like marginBottom will be a common prop with common values - planning to take a look into it and come up with some common approach for it

levadadenys replied:
Upon further look, a main differences between this PR and PRs you've mentioned are:
* (general, re all comments) that this one is for `apps/marketing`, and yours are for `packages/component-library`

* (re this specific comment) marginBottom here turns out to be a pattern across multiple conentful components, padding and margin props from your PRs are so far not a pattern, but custom/specific props for two separate component.

That being said, I agree we should follow one norm across multiple packages and code-dot-org in general. It's just that at the same time these questions should be treated with consistency, but also with some level of separation

kelbyhawn commented on frontend/apps/marketing/src/components/link/LinkContentfulDefinition.ts:
```diff
@@ -0,0 +1,81 @@
+// Creates a definition for the Typography component to be used in Contentful Studio
```
Nit:
```suggestion
// Creates a definition for the Link component to be used in Contentful Studio
```

kelbyhawn commented on frontend/apps/marketing/src/components/link/LinkContentfulDefinition.ts:
```diff
@@ -0,0 +1,81 @@
+    description:
+      'A standalone text link that directs users to internal or external pages, with options for styling and accessibility.',
+    imageUrl:
+      'https://images.ctfassets.net/90t6bu6vlf76/3Nx2ASyu1g0rwVAPrc7Srb/9c2c3e85246d8572326b741a02b69a04/component_link_tooltip.png',
```
This is showing the Paragraph tooltip image, should be this one:

<img width="654" alt="Screenshot 2025-02-10 at 1 56 30 PM" src="https://github.com/user-attachments/assets/5c004c69-d939-483a-acce-5a861a1797fb" />

kelbyhawn commented on frontend/apps/marketing/src/components/link/__tests__/Link.test.tsx:
```diff
@@ -0,0 +1,53 @@
+    expect(screen.getByRole('link')).toHaveAttribute('href', defaultProps.href);
+  });
+
+  it('applies correct CSS classes based on props', () => {
+    render(<Link {...defaultProps} />);
+    const linkElement = screen.getByRole('link');
+    expect(linkElement).toHaveClass('link');
+    expect(linkElement).toHaveClass('link-purple');
+    expect(linkElement).toHaveClass('link-m');
+    expect(linkElement).toHaveClass('link-bottomMargin-none');
+  });
```
Do we want to test for CSS classes? I remember you saying we try not to do this [here](https://github.com/code-dot-org/code-dot-org/pull/63191#discussion_r1915449668). Or is this different since it's testing the props?

levadadenys replied:
Good question!
Reason for different approach between these two is caused by these two being two separate instances with different tools.

For `component-library` we have storybook and eyes tests, for `marketing` we currently only have jest tests. I feel like we're planning to also add eyes tests to marketing, but not sure if it'll be covering link component specifically. (It probably will, but when?)

That being said, I'm happy to remove this test, it's just that I don't see any other ways to actually automatically test if our marketing custom functionality is working correctly. We can discuss it on checkin

kelbyhawn replied:
I think this is ok to keep for now! We can figure out the testing later like we talked about in Slack.

kelbyhawn commented on frontend/apps/marketing/src/components/link/link.module.scss:
```diff
@@ -0,0 +1,53 @@
+  margin-top: 0;
+}
+
+// Currently purple color styles are match component-library styles completely so we are not editing it.
+//a.link-purple {
+//}
```
Do we need this here at all?

levadadenys replied:
It's a nice to have for the future, not a need to have.
I added this to eliminate possible questions re why we don't have .link-purple styles while passing it as a className to the prop by documenting it in this way.
Plus, in case we will need to extend `link-purple` styles in the future - all we will need to do is just uncomment this and start writing the actual styles, without worrying that these styles will be in the correct place

levadadenys replied:
Small addition for context: if it was in `packages/component-library` - we would not have this problem since we control/adjust everything there, meanwhile in `apps/marketing` we rather make adjustments to existing styles and/or adding new custom styles, meaning some degree of documenting might be really helpful for the people of future.

Good part here is that I think once we agree on something here, we may follow it for possible future cases

kelbyhawn replied:
Ah ok that makes sense! Like if marketing wants to add a pink button we can do that here? It might help to add a comment explaining that like `Leaving this here in case we want to add additional color options to the marketing sites.`

levadadenys replied:
Ok, I removed this comment since it raises more questions then answers and looks like color logic will be refactored pretty soon anyways.

> Like if marketing wants to add a pink button we can do that here? It might help to add a comment explaining that like Leaving this here in case we want to add additional color options to the marketing sites.

Not exactly. It was there just to show that .link-purple className actually exists and it's empty because Currently purple color styles are match component-library styles completely so we are not editing it. (Specifically what was commented there)
In order to add more colors support - just adding new className won't be enough, since there're much more then this one step. (definitions,props,etc)

kelbyhawn commented on frontend/apps/marketing/src/components/link/link.module.scss:
```diff
@@ -0,0 +1,53 @@
+  }
+}
+
+a.link-l {
```
Nit: can you add the comment structure here like we have in the other components? (And `// Link margin` on line 39)
```suggestion
// Link sizes
a.link-l {
```

kelbyhawn commented on frontend/apps/marketing/src/components/link/LinkContentfulDefinition.ts:
```diff
@@ -0,0 +1,81 @@
+
+export const LinkContentfulComponentDefinition: ComponentDefinition = {
+  id: 'link',
+  name: 'Link',
```
Nit: rename this to "Text Link" — this should match what Moshe had in the Figma, and trying to reduce any confusion for editors.

kelbyhawn reviewed (commented):
Overall looks great! 🎉 Left a few comments and nits.

levadadenys commented:
Updated PR with review fixes and some general refactors/improvements. Here're couple videos illustrating current visual state of Link component in cms

https://github.com/user-attachments/assets/f9870bfa-682b-431d-8082-b7138c565dcb


https://github.com/user-attachments/assets/bbccfa1a-625e-480c-a16e-514920d1e036

stephenliang commented on frontend/apps/marketing/src/components/common/definitions/index.ts:
```diff
@@ -0,0 +1,24 @@
+  group: 'style',
+  validations: {
+    in: [
+      {value: 'none', displayName: '0px (None)'},
+      {value: 'xs', displayName: '8px (Extra Small)'},
+      {value: 's', displayName: '16px (Small)'},
+      {value: 'm', displayName: '24px (Medium)'},
```
Just a nit - I don't think we need to include the exact sizes in that dropdown, just XS, S, M would be sufficient.

levadadenys replied:
I think we should, at least for now, and then update later if needed based on feedback. Without exact sizes it's too abstract

stephenliang replied:
I'm not sure that the target users of this would be looking at the pixel count, What do you think @moshebaricdo ? 

That being said, if we want to preserve pixels in the text then we should at least make the dropdown sourced the SCSS so they don't deviate (if we end up changing the pixels in the mixin).

moshebaricdo replied:
I think we just want to be consistent. I don't have a strong preference, but the divider and section which are live in Contentful just use "Large," "Medium," etc without a pixel value. I think that might be a good thing since different components have different pixel values, which is intentional but may be confusing for users.

kelbyhawn replied:
I agree that we don't need the pixel values, but can always revisit later since this is a quick fix.

stephenliang commented on frontend/apps/marketing/src/components/link/Link.tsx:
```diff
@@ -0,0 +1,52 @@
+  /** Link URL */
+  href: string;
+  /** Link color */
+  color: 'purple' | 'white';
```
Per the theming discussion yesterday, would this be `'primary' | 'secondary'` or rather, when would the link be purple vs white and would the logic be automated by its parent container setting the appropriate `data-theme`?

levadadenys replied:
That's definitely out of scope of this ticket and worth revisiting later when theming will be implemented. It's possible that we will switch to `'primary' | 'secondary'` later on, but we need to implement theme switching first and discuss whole theme switching plan

stephenliang replied:
Just to note, it is fine to change things now as we do not have users yet (for at least a couple more weeks). However, it will become quite hard to change once we start having content due to data incompatibility. For example, we'll have content records and experiences on Contentful that map to "purple", and if we change "purple" to "primary" then we'll need to support both "purple" and "primary" for some time until all content records that reference "purple" are changed.

We will want to have our variables and props sorted prior to building pages or we'll need to support old and new props and do a data migration.

moshebaricdo replied:
That's a good point, I would like to move in the direction of theming sooner than later. In the case of the link, we would just need "primary," since a white link would be determined by the section's theme. So really we wouldn't even need a color option...

levadadenys replied:
Totally agree! Pretty reasonable, this should encourage us to implement theme swithching asap and arrange all color props to follow one principle

cc @moshebaricdo as we should also cover this in figma, if possible, this or begining of next sprint based on @stephenliang 's comment

moshebaricdo replied:
Got it! I'll start going through Figma and make sure props and option sets are updated to match the theming approach!

kelbyhawn replied:
Does this work need a ticket or some record of it somewhere?

levadadenys replied:
I'll just remove colors from this pr tomorrow. Once theming is implement it should work out of the box. Only ticket may be needed - is to actually check if that will be working once everything is ready 😁

stephenliang commented on frontend/apps/marketing/src/components/link/Link.tsx:
```diff
@@ -0,0 +1,52 @@
+  /** Link size */
+  size: ComponentSizeXSToL;
+  /** Whether Link is for internal code.org pages, or external web page. (external links are opened in new tab) */
+  isLinkExternal: boolean;
```
Do we have a task created to automate this logic? Would be good to add a jira to locations where this exists so we know all the updates that are needed.

levadadenys replied:
We don't have a separate ticket for this yet, if you have time for this - please feel free to do or code this. For now I keep everything in my head so feel free to ask. (Basically right now it's used only in 2 places - Link and Button components, and it looks like at least in this sprint there won't be any additional usages of isLinkExternal prop or anything like that). Also come discussions are saved here - https://github.com/code-dot-org/code-dot-org/pull/63558.  Last comment - https://github.com/code-dot-org/code-dot-org/pull/63558#discussion_r1946887560

This PR blocks other higher priority tickets that I have in this sprint, so I hope this PR review to be fast

kelbyhawn commented on frontend/apps/marketing/src/components/common/styles/mixins.scss:
```diff
@@ -0,0 +1,19 @@
+}
+
+@mixin common-margin-bottom-xs {
+  margin-bottom: 8px;
```
Should these be in `rems`? I'm using rems on the Divider and Section.

kelbyhawn commented on frontend/apps/marketing/src/components/common/types/index.ts:
```diff
@@ -0,0 +1,2 @@
+export type MarginBottomNoneToM = 'none' | 'xs' | 's' | 'm';
+export type MarginBottomNoneToL = MarginBottomNoneToM | 'l';
```
What do you think of making these applicable to margin and padding and call it `SpacingNoneToM`? I have padding on the Divider and Section so we can either make another set of these to be padding specific or make this less specific so we can use it on both.

levadadenys replied:
Nice idea! Yeah, we can do that, plus move this types to component-library package then

kelbyhawn replied:
Do we need a ticket for this?

levadadenys replied:
I already implemented this :)

kelbyhawn replied:
Ah amazing!! 🎉

kelbyhawn commented on frontend/apps/marketing/src/components/link/LinkContentfulDefinition.ts:
```diff
@@ -0,0 +1,69 @@
+    href: {
+      displayName: 'Link URL',
+      type: 'Text',
+      defaultValue: 'https://code.org',
```
I think this (`defaultValue: 'https://code.org',`) can be removed since the url will come from a content entry.

kelbyhawn commented on frontend/apps/marketing/src/components/link/Link.tsx:
```diff
@@ -0,0 +1,52 @@
+    size={size}
+    className={classNames(
+      moduleStyles.link,
+      moduleStyles[`link-${color}`],
+      moduleStyles[`link-${size}`],
```
Based on [this comment](https://github.com/code-dot-org/code-dot-org/pull/63824#discussion_r1951174904), should these be `link-color-` and `link-size-`?

levadadenys replied:
I'm sorry, something went wrong and part of that comment is missing. :( 
Specifically this part (bold):
I figured out that while component-{size}, component-{color} **works fine(naming is pretty straight-forward and selft explanatory)**, for props like marginBottom it feels more intuitive/easier to read later to use component-marginBottom-${marginBottom} className structure.

I'll also updated the comment you've shared just in case.

Now answering your question - I think color and size might be two props that may be fine with current naming. Though I start having some doubts on it now, all of dsco package components now follow these structure, meaning it might be worth updating everything at once if we decided to also name these link-color-color and link-size-size

kelbyhawn replied:
Closing the loop: discussed offline, ticket is here to update existing DSCO components to use this [CMS-352](https://codedotorg.atlassian.net/browse/CMS-352)

kelbyhawn commented on frontend/apps/marketing/src/components/link/LinkContentfulDefinition.ts:
```diff
@@ -0,0 +1,69 @@
+      group: 'content',
+    },
+    children: {
+      displayName: 'Content',
```
What is this for? 

<img width="261" alt="Screenshot 2025-02-12 at 9 01 47 AM" src="https://github.com/user-attachments/assets/50f60625-5ce3-46d4-97a9-5a6b5630533e" />

levadadenys replied:
link text/text displayed in the link/visible link text/link label

kelbyhawn replied:
Ah got it! Can we rename this to one of those, maybe "Link text label"? cc @moshebaricdo thoughts here? If y'all think it's fine we can keep it as "Content" but if I'm confused it might confuse others too.

kelbyhawn replied:
(This also isn't blocking and can be a polish followup.)

moshebaricdo replied:
+1 We want this to be "Link Label" and "Link URL"

kelbyhawn reviewed (commented):
Ok just a few more things, but this is so close!

levadadenys commented:
> Very exciting! Just want to confirm default size for link is M (like with paragraph).

That's correct, default size for link is M

kelbyhawn reviewed (approved):
Woo hoo! 🎉 🚀 Approving now but agree with Moshe on updating the "Content" label.

levadadenys commented:
Fixed all the comments except for [this one](https://github.com/code-dot-org/code-dot-org/pull/63829#discussion_r1952837531), we should address it separately.

Updated the pr, here's how Link component look now:

https://github.com/user-attachments/assets/d59639e7-af12-4b9e-a4bd-911938d501c1
